signer/src/storage/postgres.rs

```diff
@@ -446,68 +488,120 @@ impl PgStore {
+        //
+        // [1]: The sweep transaction that occurred more than
+        //      MAX_REORG_BLOCK_COUNT blocks ago may have been confirmed
+        //      more than once. If this is the case, we want the min height
+        //      of all of them.
+
+        // Given the utxo candidate above, this is our best guess of the
+        // minimum UTXO height. It might be wrong, we'll find out shortly.
+        let min_block_height_candidate = utxo_candidate
+            .block_height
+            .saturating_sub(MAX_REORG_BLOCK_COUNT);
+
+        // We want to go back at least MAX_REORG_BLOCK_COUNT blocks worth
+        // of transactions. The number here is the maximum number of
+        // transactions that the signers could get confirmed in
+        // MAX_REORG_BLOCK_COUNT bitcoin blocks, plus one. We add the one
+        // because we want the transaction right after
+        // MAX_REORG_BLOCK_COUNT worth of transactions.
+        let max_transactions = MAX_TX_PER_BITCOIN_BLOCK * MAX_REORG_BLOCK_COUNT + 1;
+
+        // Find the block height of the sweep transaction that occurred at
+        // or before block "best candidate block height" minus
+        // MAX_REORG_BLOCK_COUNT.
         //
         // We do this because the block that confirmed the UTXO with max
         // height need not be the signers' UTXO; it does not need to be on
-        // the best blockchain. So we look at the height of the block that
-        // confirmed our previous sweep transaction. This might be more
-        // than MAX_REORG_BLOCK_COUNT blocks ago, and if so, we need to get
-        // it's height. If our second-best candidate is more than
-        // MAX_REORG_BLOCK_COUNT blocks before our "best candidate", then
-        // this height is the true minimum UTXO height.
+        // the best blockchain. But if we go back at least
+        // `MAX_REORG_BLOCK_COUNT` bitcoin blocks then that UTXO is assumed
+        // to still be confirmed.
         let prev_confirmed_height_candidate = sqlx::query_scalar::<_, i64>(
             r#"
-            SELECT bb.block_height
-            FROM sbtc_signer.bitcoin_tx_inputs
-            JOIN sbtc_signer.bitcoin_transactions AS bt USING (txid)
-            JOIN sbtc_signer.bitcoin_blocks AS bb USING (block_hash)
-            WHERE prevout_type = 'signers_input'
-              AND bb.block_height < $1
-            ORDER BY bb.block_height DESC
+            WITH RECURSIVE signer_inputs AS (
+                SELECT
+                    bti.txid
+                  , bti.prevout_txid
+                  , MIN(bb.block_height) AS block_height
+                FROM sbtc_signer.bitcoin_tx_inputs AS bti
+                JOIN sbtc_signer.bitcoin_transactions USING (txid)
+                JOIN sbtc_signer.bitcoin_blocks AS bb USING (block_hash)
+                WHERE bti.prevout_type = 'signers_input'
+                  AND bb.block_height <= $1
+                GROUP BY bti.txid, bti.prevout_txid
+                LIMIT $2
```

Comment:
I'm not sure about `LIMIT $2` (`max_transactions`)  here: we may have unrelated (wrt the tx chain we want to walk, eg in case of forks) txs in here that "consume" that limit, and in the worst case scenario where we have actually need to walk `max_transactions` to find one with the block height less than the threshold we may not have all of them in `signer_inputs`.

Reply:
Yup, we spoke offline, and this is still an issue.